go.mod

```diff
@@ -1,28 +1,29 @@
 module github.com/mholt/caddy-l4
 
-go 1.21
+go 1.22.0
 
 require (
-	github.com/caddyserver/caddy/v2 v2.7.6
+	github.com/caddyserver/caddy/v2 v2.7.6-0.20240419194313-d00824f4a648
```

Comment:
I think we should keep it as-is, or even downgrade it. See: https://twitter.com/MohammedSahaf/status/1777509753616265681

Reply:
You are making a good point here, I am with you. However, the question raised remains unresolved. And I still see it necessary to keep the dependencies up to date ON THE MASTER BRANCH. Following are my points:
1. It is easier to make minor upgrades during the time, rather than solve the problem only when it is a pain in the ass
2. This project compiles well with the Caddy v2.7.6, it is only when the commit made to the upstream a month ago breaks the compilation. Since we told xcaddy to compile caddy-l4 with the master branch of Caddy, the dependencies of the latest Caddy are always pulled no matter how we pinned it in the go.mod file of this project.
3. Any pipeline triggered now would likely fail, preventing other PR from having their CI check passed. Action is needed, and pinning/downgrading the caddy in the go.mod file does not seem to work.

Comes to your hot take, it really got me when you pointed out that the division of responsibilities is required for projects maintaining executable files and libraries. And I can see you face the same problem in the relationships between Caddy and its plugins.
Since that is an unsolvable question, I would like to sort out how I think this project has been maintained, and what is missing.

There was a time when I found any attempt to upgrade Caddy would lead to some plugins breaks, due to not catching up with the upstream changes. It turns out to be difficult to locate the changes introduced several months ago (or even years ago), and then come up with a solution.
Also, The PR was merging without a CI check then, while there was a fundamental bug causing the plugin to fail to compile.

That is when I proposed introducing a CI pipeline to this project. By solving the above problems, I take the master branch to be only guaranteed compatible with the master branch of Caddy. And of course, none of the other plugins is taken into testing.

In other words, IMO, the master branch is mainly a series of snapshots that reflect what configurations, at what time, the caddy-l4 can work the latest commit of Caddy then.
Does this idea make sense to you? If not, I think we should have a conversation about what the master branch means, and how we gonna maintain it.

And what if someone is trying to build caddy-l4 with other plugins? What is a practical way we can provide this project to the end-user? I had the same worry about those problems.
The solution I proposed was to maintain a release branch with regularly releasing tags. ( https://github.com/mholt/caddy-l4/pull/130#issue-1823796189 )
IMO we need another branch with a different xcaddy setting, and it must aligned with the caddy version we pinned in the go.mod file, otherwise, the latter one is rendered useless.
The fault should be mine for not actively defending that idea and maintaining effective communication.

Reply:
> And I still see it necessary to keep the dependencies up to date ON THE MASTER BRANCH

It's not how the Go ecosystem works. This breaks more than it resolves.

> 2\. This project compiles well with the Caddy v2.7.6, it is only when the commit made to the upstream a month ago breaks the compilation.

Because it works with v2.7.6, there's no need to upgrade Caddy version in caddy-l4's go.mod file. The way Go module resolution works is different than other languages. The file specifies the minimum required version. 

> Since we told xcaddy to compile caddy-l4 with the master branch of Caddy, the dependencies of the latest Caddy are always pulled no matter how we pinned it in the go.mod file of this project.

But this is not true. The deps are merged, and then Go find the minimum version of the common dependencies that meet both requirements. If Caddy depends on v1.5.0 of dep A, and caddy-l4 requires v1.9.0 of dep A, then Go (xcaddy) will pull v1.9.0 because it's the minimum that meets both. Go doesn't pull only Caddy's deps requirements just because it's Caddy. The way xcaddy works is by creating a new Go project from scratch that puts both Caddy and caddy-l4 as direct dependencies. In that final build, Caddy doesn't have a special position besides being the plugin ecosystem provider. The Go compiler doesn't give it special consideration just because it's Caddy.

> There was a time when I found any attempt to upgrade Caddy would lead to some plugins breaks, due to not catching up with the upstream changes. It turns out to be difficult to locate the changes introduced several months ago (or even years ago), and then come up with a solution.
> Also, The PR was merging without a CI check then, while there was a fundamental bug causing the plugin to fail to compile.
> That is when I proposed introducing a CI pipeline to this project.

Fair enough, and the CI check was sorely missed. It's appreciated. But...

> I take the master branch to be only guaranteed compatible with the master branch of Caddy. In other words, IMO, the master branch is mainly a series of snapshots that reflect what configurations, at what time, the caddy-l4 can work the latest commit of Caddy then.

This is where things go wrong. This guarantee is not accomplished by continuously upgrading Caddy in the go.mod here. It should be via nightly (or some other trigger when Caddy changes) CI runs against Caddy master/HEAD. The purpose of the master branch here is NOT to be a check on Caddy. It's concerned only with itself. 



> And what if someone is trying to build caddy-l4 with other plugins? What is a practical way we can provide this project to the end-user?
> 
> But IMO, we still need another branch with a different xcaddy setting, and IT MUST ALIGNED WITH the caddy version we pinned in the go.mod file, otherwise, the latter one is rendered useless.

I'm not sure I follow where you're going with here. No other branch is needed. The only alignment needed with Caddy is IF a feature of Caddy is used here and Caddy breaks the feature. In that case, we _might_ have to upgrade the minimum required Caddy version here.

Let me explain the issue with this PR differently... If we go with this PR, this command **fails**:
```
xcaddy build v2.7.6 --with github.com/mholt/caddy-l4
```

In fact, I'm pretty sure we can reduce the Caddy version in caddy-l4's go.mod down to `v2.6.2`, and the plugin will still function will and be built with all versions of Caddy from v2.6.2 to `master`. If we reduce the Caddy version, all the below commands will work successfully:

```
xcaddy build v2.6.2 --with github.com/mholt/caddy-l4
xcaddy build v2.6.3 --with github.com/mholt/caddy-l4
xcaddy build v2.6.4 --with github.com/mholt/caddy-l4
xcaddy build v2.7.0 --with github.com/mholt/caddy-l4
xcaddy build v2.7.1 --with github.com/mholt/caddy-l4
xcaddy build v2.7.2 --with github.com/mholt/caddy-l4
xcaddy build v2.7.3 --with github.com/mholt/caddy-l4
xcaddy build v2.7.4 --with github.com/mholt/caddy-l4
xcaddy build v2.7.5 --with github.com/mholt/caddy-l4
xcaddy build v2.7.6 --with github.com/mholt/caddy-l4
xcaddy build master --with github.com/mholt/caddy-l4
```

Otherwise, we break the users for no benefit.